internal/provider/kubernetes/controller.go

```diff
@@ -782,6 +769,56 @@ func (r *gatewayAPIReconciler) findReferenceGrant(ctx context.Context, from, to
 	return nil, nil
 }
 
+// processFinalizers encapsulates logic for managing finalizers on GatewayClass and EnvoyProxy objects.
+func (r *gatewayAPIReconciler) processFinalizers(ctx context.Context, managedGC *gwapiv1.GatewayClass, resourceTree *gatewayapi.Resources) error {
```

Comment:
done some test, found a behaviour seems like a bug to me:

1. running `kubectl apply -f examples/kubernetes/merged-gateways.yaml`
2. `kubectl delete envoyproxy -n envoy-gateway-system custom-proxy-config` will hold
3. `kubectl delete gc merged-eg` will hold
4. `kubectl delete gtw merged-eg-1 merged-eg-2 merged-eg-3`, all gtws are removed successfully, and step 3 return successfully, but step 2 still hold. By inspecting the finializers of `custom-proxy-config`, the finializer `gatewayclass-exists-finalizer.gateway.envoyproxy.github.com` still exist.

Reply:
It seems like that's correct, the issue is that when the `GatewayClass` gets `deletionTimestamp`, we don't process `EnvoyProxy`, hence processed `resourceTree.EnvoyGatewayForGatewayClass` will be `nil`. (it also cause mergeGateways to be set to nil).
I can also try to fix this in this PR, as this makes the finalizers more complicated.